File: NET_project/Benchmarks/PixarRaytracer.cs
using System.Numerics;
using System.Runtime.CompilerServices;

namespace NET_project.Benchmarks;

public enum PixarRayHit {
	None = 0,
	Letter = 1,
	Wall = 2,
	Sun = 3
}

public unsafe struct PixarRaytracerNET : IJob
{
	public uint width;
	public uint height;
	public uint samples;
	public float result;

	public void Run() {
		result = PixarRaytracer(width, height, samples);
	}

	private static uint marsagliaZ;
	private static uint marsagliaW;

	private static float PixarRaytracer(uint width, uint height, uint samples)
	{
		marsagliaZ = 666;
		marsagliaW = 999;

		Vector3 position = new Vector3(-22.0f, 5.0f, 25.0f);
		Vector3 goal = new Vector3(-3.0f, 4.0f, 0.0f);

		goal = Vector3.Normalize(goal) - position;

		Vector3 left = new Vector3(goal.Z, 0, goal.X);

		left = Vector3.Normalize(left) / width;

		Vector3 up = Vector3.Cross(goal, left);
		Vector3 color = Vector3.Zero;
		Vector3 adjust;

		for (uint y = height; y > 0; y--) {
			for (uint x = width; x > 0; x--) {
				for (uint p = samples; p > 0; p--) {
					color += Trace(position, Vector3.Normalize((goal + left) * (x - (width / 2) + Random())) + (up * (y - (height / 2) + Random())));
				}

				color *= (1.0f / samples) + (14.0f / 241.0f);
				adjust = color + Vector3.One;
				color /= adjust;

				color *= 255.0f;
			}
		}

		return color.X + color.Y + color.Z;
	}

	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	private static float Min(float left, float right) {
		return left < right ? left : right;
	}

	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	private static float BoxTest(Vector3 position, Vector3 lowerLeft, Vector3 upperRight) {
		lowerLeft = -(position + lowerLeft);
		upperRight = -(upperRight + position);

		Vector3 vmin = Vector3.Min(lowerLeft, upperRight);
		return -Min(Min(vmin.X, vmin.Y), vmin.Z);
	}

	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	private static float Random() {
		marsagliaZ = (36969 * (marsagliaZ & 65535)) + (marsagliaZ >> 16);
		marsagliaW = (18000 * (marsagliaW & 65535)) + (marsagliaW >> 16);

		return ((marsagliaZ << 16) + marsagliaW) * 2.0f / 10000000000.0f;
	}

	private static ReadOnlySpan<float> letters => [
		/* P */ 53, 79, 53, 95, 53, 87, 57, 87, 53, 95, 57, 95,
		/* I */ 65, 79, 69, 79, 67, 79, 67, 95, 65, 95, 69, 95,
		/* X */ 73, 79, 81, 95, 73, 95, 81, 79,
		/* A */ 85, 79, 89, 95, 89, 95, 93, 79, 87, 87, 91, 87,
		/* R */ 97, 79, 97, 95, 97, 87, 101, 87, 97, 95, 101, 95, 99, 87, 105, 79
	];

	private static float Sample(Vector3 position, int* hitType) {
		float distance = 1e9f;
		Vector3 f = position;

		f.Z = 0.0f;

		for (int i = 0; i < letters.Length; i += 4) {
			// The JIT isn't quite smart enough to elide the bounds
			// checks and so we'll help it out a little bit instead.
			ref float letter = ref Unsafe.AsRef(in letters[i]);

			Vector3 begin = new Vector3((Unsafe.As<float, Vector2>(ref letter) - new Vector2(79.0f)) * 0.5f, 0.0f);
			Vector3 e = new Vector3((Unsafe.As<float, Vector2>(ref Unsafe.Add(ref letter, 2)) - new Vector2(79.0f)) * 0.5f, 0.0f) - begin;
			Vector3 o = -(f + ((begin + e) * Min(-Min(Vector3.Dot(-(begin + f), e) / e.LengthSquared(), 0.0f), 1.0f)));

			distance = Min(distance, o.LengthSquared());
		}

		distance = float.Sqrt(distance);

		Span<Vector3> curves = [
			new Vector3(11.0f, 6.0f, 0.0f),
			new Vector3(-11.0f, 6.0f, 0.0f),
		];

		for (int i = 0; i < curves.Length; i++) {
			Vector3 o = f - curves[i];
			float m = 0.0f;

			if (o.X > 0.0f) {
				m = float.Abs(o.Length() - 2.0f);
			} else {
				if (o.Y > 0.0f)
					o.Y += -2.0f;
				else
					o.Y += 2.0f;

				o.Y += o.Length();
			}

			distance = Min(distance, m);
		}

		distance = float.Pow(float.Pow(distance, 8.0f) + float.Pow(position.Z, 8.0f), 0.125f) - 0.5f;
		*hitType = (int)PixarRayHit.Letter;

		float roomDistance = Min(-Min(BoxTest(position, new Vector3(-30.0f, -0.5f, -30.0f), new Vector3(30.0f, 18.0f, 30.0f)), BoxTest(position, new Vector3(-25.0f, -17.5f, -25.0f), new Vector3(25.0f, 20.0f, 25.0f))), BoxTest(new Vector3(float.Abs(position.X) % 8, position.Y, position.Z), new Vector3(1.5f, 18.5f, -25.0f), new Vector3(6.5f, 20.0f, 25.0f)));

		if (roomDistance < distance) {
			distance = roomDistance;
			*hitType = (int)PixarRayHit.Wall;
		}

		float sun = 19.9f - position.Y;

		if (sun < distance) {
			distance = sun;
			*hitType = (int)PixarRayHit.Sun;
		}

		return distance;
	}

	private static int RayMarching(Vector3 origin, Vector3 direction, Vector3* hitPosition, Vector3* hitNormal) {
		int hitType = (int)PixarRayHit.None;
		int noHitCount = 0;
		float distance = 0.0f;

		for (float i = 0; i < 100; i += distance) {
			*hitPosition = (origin + direction) * i;
			distance = Sample(*hitPosition, &hitType);

			if (distance < 0.01f || ++noHitCount > 99) {
				*hitNormal = Vector3.Normalize(new Vector3(Sample(*hitPosition + new Vector3(0.01f, 0.0f, 0.0f), &noHitCount) - distance, Sample(*hitPosition + new Vector3(0.0f, 0.01f, 0.0f), &noHitCount) - distance, Sample(*hitPosition + new Vector3(0.0f, 0.0f, 0.01f), &noHitCount) - distance));

				return hitType;
			}
		}

		return (int)PixarRayHit.None;
	}

	private static Vector3 Trace(Vector3 origin, Vector3 direction) {
		Vector3 sampledPosition = Vector3.One;
		Vector3 normal = Vector3.One;
		Vector3 color = Vector3.One;
		Vector3 attenuation = Vector3.One;
		Vector3 lightDirection = Vector3.Normalize(new Vector3(0.6f, 0.6f, 1.0f));

		for (int bounce = 3; bounce > 0; bounce--) {
			PixarRayHit hitType = (PixarRayHit)RayMarching(origin, direction, &sampledPosition, &normal);

			switch (hitType) {
				case PixarRayHit.None:
					break;

				case PixarRayHit.Letter:
				{
					direction = (direction + normal) * (Vector3.Dot(normal, direction) * -2.0f);
					origin = (sampledPosition + direction) * 0.1f;
					attenuation *= 0.2f;

					break;
				}

				case PixarRayHit.Wall:
				{
					float incidence = Vector3.Dot(normal, lightDirection);
					float p = 6.283185f * Random();
					float c = Random();
					float s = float.Sqrt(1.0f - c);
					float g = (normal.Z < 0) ? -1.0f : 1.0f;
					float u = -1.0f / (g + normal.Z);
					float v = normal.X * normal.Y * u;

					direction = new Vector3(v, g + (normal.Y * normal.Y * u), -normal.Y * (float.Cos(p) * s)) + new Vector3(1.0f + (g * normal.X * normal.X * u), g * v, -g * normal.X) + (normal * float.Sqrt(c));
					origin = (sampledPosition + direction) * 0.1f;
					attenuation *= 0.2f;

					if (incidence > 0 && RayMarching((sampledPosition + normal) * 0.1f, lightDirection, &sampledPosition, &normal) == (int)PixarRayHit.Sun)
						color = (color + attenuation) * new Vector3(500.0f, 400.0f, 100.0f) * incidence;

					break;
				}

				case PixarRayHit.Sun:
				{
					color = (color + attenuation) * new Vector3(50.0f, 80.0f, 100.0f);
					return color;
				}
			}
		}

		return color;
	}
}

internal unsafe struct PixarRaytracerGCC : IJob {
	public uint width;
	public uint height;
	public uint samples;
	public float result;
	public string libName;

	public void Run() {
		switch (libName)
		{
			case "GCC":
				result = GccNativeBindings.benchmark_pixar_raytracer(width, height, samples);
				break;

			case "Clang":
				result = ClangNativeBindings.benchmark_pixar_raytracer(width, height, samples);
				break;

			case "MS":
				result = MsNativeBindings.benchmark_pixar_raytracer(width, height, samples);
				break;

			case "Rust":
				result = RustNativeBindings.benchmark_pixar_raytracer(width, height, samples);
				break;
		}
	}
}

File: rust_lib/src/f32_benchmarks/PixarRaytracer.rs


#[derive(Clone, Copy)]
struct Vector
{
    pub(crate) x: f32,
    pub(crate) y: f32,
    pub(crate) z: f32,
}

enum PixarRayHit
{
    PIXAR_RAYTRACER_NONE = 0,
    PIXAR_RAYTRACER_LETTER = 1,
    PIXAR_RAYTRACER_WALL = 2,
    PIXAR_RAYTRACER_SUN = 3
}


fn benchmark_pixar_raytracer_add(mut left: Vector, right: Vector) -> Vector
{
    left.x += right.x;
    left.y += right.y;
    left.z += right.z;

    return left;
}

fn benchmark_pixar_raytracer_add_float(mut vector: Vector, value: f32) -> Vector
{
    vector.x += value;
    vector.y += value;
    vector.z += value;

    return vector;
}

fn benchmark_pixar_raytracer_cross(to: Vector, from: Vector) -> Vector
{
    let mut vector: Vector = Vector
    {
        x: 0.0,
        y: 0.0,
        z: 0.0
    };

    vector.x = to.y * from.z - to.z * from.y;
    vector.y = to.z * from.x - to.x * from.z;
    vector.z = to.x * from.y - to.y * from.x;

    return vector;
}

fn benchmark_pixar_raytracer_inverse(vector: Vector) -> Vector
{
    return benchmark_pixar_raytracer_multiply_float(vector, 1.0 / f32::sqrt(benchmark_pixar_raytracer_modulus_self(vector)));
}

fn benchmark_pixar_raytracer_modulus(left: Vector, right: Vector) -> f32
{
    return left.x * right.x + left.y * right.y + left.z * right.z;
}

fn benchmark_pixar_raytracer_modulus_self(vector: Vector) -> f32
{
    return vector.x * vector.x + vector.y * vector.y + vector.z * vector.z;
}

fn benchmark_pixar_raytracer_multiply(mut left: Vector, right: Vector) -> Vector
{
    left.x *= right.x;
    left.y *= right.y;
    left.z *= right.z;

    return left;
}

fn benchmark_pixar_raytracer_multiply_float(mut vector: Vector, value: f32) -> Vector
{
    vector.x *= value;
    vector.y *= value;
    vector.z *= value;

    return vector;
}

fn benchmark_pixar_raytracer_min(left: f32, right: f32) -> f32
{
    return if (left < right)
    {
        left
    }
    else
    {
        right
    }
}

fn benchmark_pixar_raytracer_box_test(position: Vector, mut lowerLeft: Vector, mut upperRight: Vector) -> f32
{
    lowerLeft = benchmark_pixar_raytracer_multiply_float(benchmark_pixar_raytracer_add(position, lowerLeft), -1f32);
    upperRight = benchmark_pixar_raytracer_multiply_float(benchmark_pixar_raytracer_add(upperRight, position), -1f32);

    return -benchmark_pixar_raytracer_min(
        benchmark_pixar_raytracer_min(
            benchmark_pixar_raytracer_min(lowerLeft.x, upperRight.x),
            benchmark_pixar_raytracer_min(lowerLeft.y, upperRight.y)
        ),
        benchmark_pixar_raytracer_min(lowerLeft.z, upperRight.z)
    );
}

fn benchmark_pixar_raytracer_ray_marching(origin: Vector, direction: Vector, hitPosition: &mut Vector, hitNormal: &mut Vector) -> i32
{
    let mut hitType: i32 = PixarRayHit::PIXAR_RAYTRACER_NONE as i32;
    let mut noHitCount: i32 = 0;
    let mut distance: f32;

    let mut i: f32 = 0.0;
    while (i < 100.0)
    {
        *hitPosition = benchmark_pixar_raytracer_multiply_float(benchmark_pixar_raytracer_add(origin, direction), i);
        distance = benchmark_pixar_raytracer_sample(*hitPosition, &mut hitType);

        noHitCount += 1;
        if (distance < 0.01 || noHitCount > 99)
        {
            *hitNormal = benchmark_pixar_raytracer_inverse(
                Vector {
                    x: benchmark_pixar_raytracer_sample(benchmark_pixar_raytracer_add(*hitPosition, Vector { x: 0.01, y: 0.0, z: 0.0 }), &mut noHitCount) - distance,
                    y: benchmark_pixar_raytracer_sample(benchmark_pixar_raytracer_add(*hitPosition, Vector { x: 0.0, y: 0.01, z: 0.0 }), &mut noHitCount) - distance,
                    z: benchmark_pixar_raytracer_sample(benchmark_pixar_raytracer_add(*hitPosition, Vector { x: 0.0, y: 0.0, z: 0.01 }), &mut noHitCount) - distance
                }
            );

            return hitType;
        }

        i += distance;
    }

    return PixarRayHit::PIXAR_RAYTRACER_NONE as i32;
}

fn benchmark_pixar_raytracer_random(marsagliaZ: &mut u32, marsagliaW: &mut u32) -> f32
{
    *marsagliaZ = 36969 * (*marsagliaZ & 65535) + (*marsagliaZ >> 16);
    *marsagliaW = 18000 * (*marsagliaW & 65535) + (*marsagliaW >> 16);

    return (((*marsagliaZ << 16) + *marsagliaW) as f32) * 2.0 / 10000000000.0;
}

fn benchmark_pixar_raytracer_sample(position: Vector, hitType: &mut i32) -> f32
{
    const size: usize = 60;

    let mut distance: f32 = 1e9;
    let mut f: Vector = position;
    let mut letters: [u8; size] = [0; size];

    // P              // I              // X              // A              // R
    letters[0]  = 53; letters[12] = 65; letters[24] = 73; letters[32] = 85; letters[44] = 97; letters[56] = 99;
    letters[1]  = 79; letters[13] = 79; letters[25] = 79; letters[33] = 79; letters[45] = 79; letters[57] = 87;
    letters[2]  = 53; letters[14] = 69; letters[26] = 81; letters[34] = 89; letters[46] = 97; letters[58] = 105;
    letters[3]  = 95; letters[15] = 79; letters[27] = 95; letters[35] = 95; letters[47] = 95; letters[59] = 79;
    letters[4]  = 53; letters[16] = 67; letters[28] = 73; letters[36] = 89; letters[48] = 97;
    letters[5]  = 87; letters[17] = 79; letters[29] = 95; letters[37] = 95; letters[49] = 87;
    letters[6]  = 57; letters[18] = 67; letters[30] = 81; letters[38] = 93; letters[50] = 101;
    letters[7]  = 87; letters[19] = 95; letters[31] = 79; letters[39] = 79; letters[51] = 87;
    letters[8]  = 53; letters[20] = 65;                   letters[40] = 87; letters[52] = 97;
    letters[9]  = 95; letters[21] = 95;                   letters[41] = 87; letters[53] = 95;
    letters[10] = 57; letters[22] = 69;                   letters[42] = 91; letters[54] = 101;
    letters[11] = 95; letters[23] = 95;                   letters[43] = 87; letters[55] = 95;

    f.z = 0.0;

    let mut i: usize = 0;
    while (i < size)
    {
        let begin: Vector = benchmark_pixar_raytracer_multiply_float(
            Vector { x: letters[i] as f32 - 79.0, y: letters[i + 1] as f32 - 79.0, z: 0.0 }, 0.5
        );
        let e: Vector = benchmark_pixar_raytracer_add(
            benchmark_pixar_raytracer_multiply_float(
                Vector { x: letters[i + 2] as f32 - 79.0, y: letters[i + 3] as f32 - 79.0, z: 0.0 }, 0.5
            ),
            benchmark_pixar_raytracer_multiply_float(begin, -1.0)
        );
        let o: Vector = benchmark_pixar_raytracer_multiply_float(
            benchmark_pixar_raytracer_add(
                f,
                benchmark_pixar_raytracer_multiply_float(
                    benchmark_pixar_raytracer_add(begin, e),
                    benchmark_pixar_raytracer_min(
                        -benchmark_pixar_raytracer_min(
                            benchmark_pixar_raytracer_modulus(
                                benchmark_pixar_raytracer_multiply_float(benchmark_pixar_raytracer_add(begin, f), -1.0),
                                e
                            ) / benchmark_pixar_raytracer_modulus_self(e),
                            0.0
                        ),
                        1.0
                    )
                )
            ),
            -1.0
        );

        distance = benchmark_pixar_raytracer_min(distance, benchmark_pixar_raytracer_modulus_self(o));

        i += 4;
    }

    distance = distance.sqrt();

    let mut curves: [Vector; 2] = [Vector {x: 0.0, y: 0.0, z: 0.0}; 2];

    curves[0] = Vector {x: -11.0, y: 6.0, z: 0.0};
    curves[1] = Vector {x:  11.0, y: 6.0, z: 0.0};

    let mut iInt: i32 = 1;
    while (iInt >= 0)
    {
        let i: usize = iInt as usize;
        let mut o: Vector = benchmark_pixar_raytracer_add(f, benchmark_pixar_raytracer_multiply_float(curves[i], -1.0));
        let mut m: f32 = 0.0;

        if (o.x > 0.0)
        {
            m = f32::abs(benchmark_pixar_raytracer_modulus_self(o).sqrt() - 2.0);
        }
        else
        {
            if (o.y > 0.0)
            {
                o.y += -2.0;
            }
            else
            {
                o.y += 2.0;
            }

            o.y += benchmark_pixar_raytracer_modulus_self(o).sqrt();
        }

        distance = benchmark_pixar_raytracer_min(distance, m);

        iInt -= 1;
    }

    distance = f32::powf(distance.powf(8.0) + position.z.powf(8.0), 0.125) - 0.5;
    *hitType = PixarRayHit::PIXAR_RAYTRACER_LETTER as i32;

    let roomDistance: f32 = benchmark_pixar_raytracer_min(
        -benchmark_pixar_raytracer_min(
            benchmark_pixar_raytracer_box_test(
                position,
                Vector { x: -30.0, y: -0.5, z: -30.0 }, Vector { x: 30.0, y: 18.0, z: 30.0 }
            ),
            benchmark_pixar_raytracer_box_test(
                position,
                Vector { x: -25.0, y: -17.5, z: -25.0 }, Vector { x: 25.0, y: 20.0, z: 25.0 }
            )
        ),
        benchmark_pixar_raytracer_box_test(
            Vector { x: position.x.abs() % 8.0, y: position.y, z: position.z },
            Vector { x: 1.5, y: 18.5, z: -25.0 },
            Vector { x: 6.5, y: 20.0, z: 25.0 }
        )
    );

    if (roomDistance < distance)
    {
        distance = roomDistance;
        *hitType = PixarRayHit::PIXAR_RAYTRACER_WALL as i32;
    }

    let sun: f32 = 19.9 - position.y;

    if (sun < distance)
    {
        distance = sun;
        *hitType = PixarRayHit::PIXAR_RAYTRACER_SUN as i32;
    }

    return distance;
}

fn benchmark_pixar_raytracer_trace(mut origin: Vector, mut direction: Vector, marsagliaZ: &mut u32, marsagliaW: &mut u32) -> Vector
{
    let mut sampledPosition: Vector = Vector { x: 1.0f32, y: 1.0f32, z: 1.0f32 };
    let mut normal:          Vector = Vector { x: 1.0f32, y: 1.0f32, z: 1.0f32 };
    let mut color:           Vector = Vector { x: 1.0f32, y: 1.0f32, z: 1.0f32 };
    let mut attenuation:     Vector = Vector { x: 1.0f32, y: 1.0f32, z: 1.0f32 };
    let lightDirection:  Vector = benchmark_pixar_raytracer_inverse(Vector { x: 0.6f32, y: 0.6f32, z: 1.0f32 });

    let mut bounce: i32 = 3;
    while (bounce > 0)
    {
        let hitType: i32 = benchmark_pixar_raytracer_ray_marching(origin, direction, &mut sampledPosition, &mut normal);

        if (hitType == PixarRayHit::PIXAR_RAYTRACER_NONE as i32)
        {
            //
        }
        else if (hitType == PixarRayHit::PIXAR_RAYTRACER_LETTER as i32)
        {
            direction = benchmark_pixar_raytracer_multiply_float(
                benchmark_pixar_raytracer_add(direction, normal),
                benchmark_pixar_raytracer_modulus(normal, direction) * -2.0
            );
            origin = benchmark_pixar_raytracer_multiply_float(benchmark_pixar_raytracer_add(sampledPosition, direction), 0.1);
            attenuation = benchmark_pixar_raytracer_multiply_float(attenuation, 0.2);
        }
        else if (hitType == PixarRayHit::PIXAR_RAYTRACER_WALL as i32)
        {
            let incidence: f32 = benchmark_pixar_raytracer_modulus(normal, lightDirection);
            let p: f32 = 6.283185 * benchmark_pixar_raytracer_random(marsagliaZ, marsagliaW);
            let c: f32 = benchmark_pixar_raytracer_random(marsagliaZ, marsagliaW);
            let s: f32 = f32::sqrt(1.0 - c);
            let g: f32 = match (normal.z < 0.0)
            {
                true => -1.0,
                false => 1.0
            };
            let u: f32 = -1.0 / (g + normal.z);
            let v: f32 = normal.x * normal.y * u;

            direction = benchmark_pixar_raytracer_add(
                benchmark_pixar_raytracer_add(
                    Vector { x: v, y: g + normal.y * normal.y * u, z: -normal.y * (p.cos() * s) },
                    Vector { x: 1.0 + g * normal.x * normal.x * u, y: g * v, z: -g * normal.x }
                ),
                benchmark_pixar_raytracer_multiply_float(normal, c.sqrt())
            );
            origin = benchmark_pixar_raytracer_multiply_float(benchmark_pixar_raytracer_add(sampledPosition, direction), 0.1);
            attenuation = benchmark_pixar_raytracer_multiply_float(attenuation, 0.2);

            if (
                incidence > 0.0 &&
                benchmark_pixar_raytracer_ray_marching(
                    benchmark_pixar_raytracer_multiply_float(benchmark_pixar_raytracer_add(sampledPosition, normal), 0.1),
                    lightDirection, &mut sampledPosition, &mut normal
                ) == (PixarRayHit::PIXAR_RAYTRACER_SUN as i32)
            )
            {
                color = benchmark_pixar_raytracer_multiply_float(
                    benchmark_pixar_raytracer_multiply(benchmark_pixar_raytracer_add(color, attenuation), Vector { x: 500.0, y: 400.0, z: 100.0 }),
                    incidence
                );
            }
        }
        else if (hitType == PixarRayHit::PIXAR_RAYTRACER_SUN as i32)
        {
            color = benchmark_pixar_raytracer_multiply(benchmark_pixar_raytracer_add(color, attenuation), Vector { x: 50.0, y: 80.0, z: 100.0 });

            break;
        }

        bounce -= 1;
    }

    return color;
}

pub(crate) fn benchmark_pixar_raytracer_main(width: u32, height: u32, samples: u32) -> f32
{
    let mut marsagliaZ: u32 = 666;
    let mut marsagliaW: u32 = 999;

    let position: Vector = Vector
    {
        x: -22.0f32,
        y: 5.0f32,
        z: 25.0f32
    };
    let mut goal: Vector = Vector
    {
        x: -3.0f32,
        y: 4.0f32,
        z: 0.0f32
    };

    goal = benchmark_pixar_raytracer_add(benchmark_pixar_raytracer_inverse(goal), benchmark_pixar_raytracer_multiply_float(position, -1.0f32));

    let mut left: Vector = Vector
    {
        x: goal.z,
        y: 0f32,
        z: goal.x,
    };

    left = benchmark_pixar_raytracer_multiply_float(benchmark_pixar_raytracer_inverse(left), 1.0 / width as f32);

    let up: Vector = benchmark_pixar_raytracer_cross(goal, left);
    let mut color: Vector = Vector
    {
        x: 0.0,
        y: 0.0,
        z: 0.0
    };
    let mut adjust: Vector;

    let mut y: u32 = height;
    while (y > 0)
    {
        let mut x: u32 = width;
        while (x > 0)
        {
            let mut p: u32 = samples;
            while (p > 0)
            {
                color = benchmark_pixar_raytracer_add(
                    color,
                    benchmark_pixar_raytracer_trace(
                        position,
                        benchmark_pixar_raytracer_add(
                            benchmark_pixar_raytracer_inverse(
                                benchmark_pixar_raytracer_multiply_float(
                                    benchmark_pixar_raytracer_add(goal, left),
                                    (x - width / 2) as f32 + benchmark_pixar_raytracer_random(&mut marsagliaZ, &mut marsagliaW)
                                )
                            ),
                            benchmark_pixar_raytracer_multiply_float(
                                up,
                                (y - height / 2) as f32 + benchmark_pixar_raytracer_random(&mut marsagliaZ, &mut marsagliaW)
                            )
                        ),
                        &mut marsagliaZ,
                        &mut marsagliaW
                    )
                );
                p -= 1;
            }

            color = benchmark_pixar_raytracer_multiply_float(color, (1.0 / samples as f32) + 14.0 / 241.0);
            adjust = benchmark_pixar_raytracer_add_float(color, 1.0);
            color = Vector
            {
                x: color.x / adjust.x,
                y: color.y / adjust.y,
                z: color.z / adjust.z
            };

            color = benchmark_pixar_raytracer_multiply_float(color, 255.0);

            x -= 1;
        }

        y -= 1;
    }

    return color.x + color.y + color.z;
}
